File: Python/refer-offer/shu_zu_zhong_de_ni_xu_dui_lcof.py
from typing import *

def reversePairs(nums: List[int]) -> int:
    temp = [0] * len(nums)
    def merge_sort(left, right):
        if left >= right:
            return 0
        mid = (left + right) // 2
        res = merge_sort(left, mid) + merge_sort(mid + 1, right)
        i, j = left, mid + 1
        temp[left:right+1] = nums[left:right+1]
        for k in range(left, right + 1):
            if i == mid + 1:
                nums[k] = temp[j]
                j += 1
            elif j == right + 1 or temp[i] <= temp[j]:
                nums[k] = temp[i]
                i += 1
            else:
                nums[k] = temp[j]
                j += 1
                res += mid - i + 1
        
        return res
    return merge_sort(0, len(nums) - 1)
    
    
            

File: Java/src/com/dbc/algorithm/ReferOffer/ShuZuZhongDeNiXuDuiLcof.java
package com.dbc.algorithm.ReferOffer;

public class ShuZuZhongDeNiXuDuiLcof {
    public int reversePairs(int[] nums) {
        return mergeSort(0, nums.length - 1, nums);
    }

    public int mergeSort(int left, int right, int[] nums) {
        if (left >= right) {
            return 0;
        }
        int mid = (left + right) / 2;
        int res = mergeSort(left, mid, nums) + mergeSort(mid + 1, right, nums);
        int[] temp = new int[right - left + 1];
        for (int i = left; i < right + 1; i++) {
            temp[i - left] = nums[i];
        }
        int point1 = left, point2 = mid + 1;
        for (int i = left; i < right + 1; i++) {
            if (point1 == mid + 1) {
                nums[i] = temp[point2 - left];
                point2++;
            } else if (point2 == right + 1 || temp[point1 - left] <= temp[point2 - left]) {
                nums[i] = temp[point1 - left];
                point1++;
            } else {
                nums[i] = temp[point2 - left];
                point2++;
                res += mid - point1 + 1;
            }
        }
        return res;

    }
}
